### PR TITLE
(Mostly) Type-check where clauses on associated types

### PR DESCRIPTION
**Pull request by huonw**
This doesn't yet check that conforming types satisfy the requirements, nor does it do basic sensible-ness checks on the requirements (like `associatedtype T where T == Int`).

### CODE REVIEW

**Comment by huonw** [comment]
Will address the missing pieces and rebase tomorrow.

**Comment by slavapestov** [comment]
'associatedtype T where T == Int' could be equivalent to just 'typealias T == Int'

File: lib/AST/SubstitutionMap.cpp [hunk 1]
```diff
@@ -226,18 +226,14 @@ SubstitutionMap
...
-      assert(origType->isEqual(protocolSelfType) &&
-             protoType->getDecl() == protocol);
-      return conformance;
-    });
+  // The substitutions required for a protocol are simple: Self -> `selfType`,
+  // and Self: `protocol`.
+  auto protocolSelfType = cast<GenericTypeParamType>(
+      protocol->getSelfInterfaceType()->getCanonicalType());
+  SubstitutionMap map;
+  map.addSubstitution(protocolSelfType, selfType);
```
Comment (slavapestov) [inline lib/AST/SubstitutionMap.cpp]: Eventually SubstitutionMaps will be arrays indexed by generic signature requirements; how will this work?

Let's discuss this tomorrow.

File: lib/AST/Type.cpp [hunk 2]
```diff
@@ -3240,12 +3240,29 @@ TypeSubstitutionMap TypeBase::getContextSubstitutions(const DeclContext *dc) {
 
 SubstitutionMap TypeBase::getContextSubstitutionMap(
   ModuleDecl *module, const DeclContext *dc) {
+  auto lookup = LookUpConformanceInModule(module);
+
+  // If we're substituting in a protocol, we can avoid the full
+  // generic-signature-based lookup, because the substitutions are easy to
+  // compute. It is important to do this, because name-lookup needs to use this,
+  // and the generic signature of a protocol may not yet be computed.
```
Comment (slavapestov) [inline lib/AST/Type.cpp]: A protocol generic signature is completely trivial though, shouldn't we be able to have one here?

File: lib/Parse/ParseDecl.cpp [hunk 3]
```diff
@@ -3236,15 +3236,36 @@ ParserResult<TypeDecl> Parser::parseDeclAssociatedType(Parser::ParseDeclOptions
...
+  // Parse a 'where' clause if present.
   if (Tok.is(tok::kw_where)) {
-    GenericParamList *unused = nullptr;
-    auto whereStatus = parseFreestandingGenericWhereClause(
-        unused, WhereClauseKind::AssociatedType);
-    if (whereStatus.shouldStopParsing())
-      return whereStatus;
+    if (Context.isSwiftVersion3()) {
+      // Swift 3 doesn't support where clauses on associated types, so let's
+      // keep exactly the same parsing behaviour as it had.
```
Comment (slavapestov) [inline lib/Parse/ParseDecl.cpp]: I think we want the new feature to still work in Swift 3 mode too. This was the decision for nested generics and generic subscripts. Only incompatible changes should be limited to Swift 4 mode.

File: lib/Sema/TypeChecker.h [hunk 4]
```diff
@@ -437,6 +437,11 @@ enum TypeResolutionFlags : unsigned {
 
   /// Whether we are checking the outermost type of a computed property setter's newValue
   TR_ImmediateSetterNewValue = 0x1000000,
+
+  /// Whether we are checking the outermost layer of types in an inheritance
+  /// clause on something other than an enum (i.e. V, but not U or W, in class
+  /// T: U.V<W>)
+  TR_NonEnumInheritanceClauseOuterLayer = 0x2000000,
```
Comment (slavapestov) [inline lib/Sema/TypeChecker.h]: We need to come up with a better name...

File: lib/AST/Type.cpp (hunk 2 above)
Comment (huonw) [inline lib/AST/Type.cpp]: I'm not sure it is trivial enough: don't all the implied requirements have to be stored too (in the `GenericSignatureBuilder`)?

**Comment by huonw** [comment]
> 'associatedtype T where T == Int' could be equivalent to just 'typealias T == Int'

Is it worth allowing it just for that purpose, rather than outlawing it and suggesting a `typealias`? (At least, once typealiases work well enough...)

**Comment by jrose-apple** [comment]
>> 'associatedtype T where T == Int' could be equivalent to just 'typealias T == Int'
>
> Is it worth allowing it just for that purpose, rather than outlawing it and suggesting a `typealias`? (At least, once typealiases work well enough...)

IIRC this is currently accepted but does not *require* that the typealias be used:

```swift
protocol Base {
  associatedtype Assoc
}
protocol Refined: Base {
  typealias Assoc = Int
}
struct Impl: Refined {
  typealias Assoc = Float
}

func test<T: Base>(_: T) -> T.Assoc? { return nil }
print(type(of: test(Impl()))) // Optional<Float>
```

So at least in Swift 3 mode the associated-type-with-where has a stronger meaning.

File: lib/AST/Type.cpp (hunk 2 above)
Comment (DougGregor) [inline lib/AST/Type.cpp]: The signature itself should be trivial `<Self: Proto>`; when we create a `GenericSignatureBuilder` we'll explode out the implied requirements so they can be queried.

To be fair, our nascent plans for storing concrete conformances into the signature will mean that `<Self: Proto>` won't always be trivial. We'll have to think about how to handle that... later.

File: lib/Parse/ParseDecl.cpp (hunk 3 above)
Comment (DougGregor) [inline lib/Parse/ParseDecl.cpp]: Generally, I think you're right. Although.. use of this feature will completely blow up IRGen, so maybe it's useful as a guard against accidental use/bug reports in the short term.

File: lib/Parse/ParseDecl.cpp [hunk 5]
```diff
@@ -3236,15 +3236,36 @@ ParserResult<TypeDecl> Parser::parseDeclAssociatedType(Parser::ParseDeclOptions
...
+        TrailingWhere =
+            TrailingWhereClause::create(Context, whereLoc, requirements);
+      } else if (whereStatus.hasCodeCompletion()) {
+        // FIXME: this is completely (hah) cargo culted.
+        if (CodeCompletion && firstTypeInComplete) {
+          CodeCompletion->completeGenericParams(nullptr);
+        } else {
+          return makeParserCodeCompletionResult<AssociatedTypeDecl>();
+        }
+      }
```
Comment (DougGregor) [inline lib/Parse/ParseDecl.cpp]: Can we simplify the Swift 3 vs. Swift 4 logic here? Always go down the parsing path, but just drop `TrailingWhere` on the floor with a "this is not implemented yet; use -swift-version 4 to experiment" error of some sort?

File: lib/Sema/GenericTypeResolver.h [hunk 6]
```diff
@@ -144,8 +144,34 @@ class GenericTypeToArchetypeResolver : public GenericTypeResolver {
...
+
+  virtual void recordParamType(ParamDecl *decl, Type ty);
+};
+
+/// Generic type resolver that only handles what can appear in a protocol
+/// definition, i.e. Self, and Self.A.B.C dependent types.
+///
+/// This should only be used when resolving/validating where clauses in
+/// protocols.
+class ProtocolRequirementTypeResolver : public GenericTypeResolver {
```
Comment (DougGregor) [inline lib/Sema/GenericTypeResolver.h]: It deeply saddens me to have another `GenericTypeResolver`, but I agree that this is the right thing now.

File: lib/AST/GenericSignatureBuilder.cpp [hunk 7]
```diff
@@ -1101,7 +1101,11 @@ bool GenericSignatureBuilder::addConformanceRequirement(PotentialArchetype *PAT,
             return true;
         }
 
-        continue;
+        if (auto WhereClause = AssocType->getTrailingWhereClause()) {
+          for (auto &req : WhereClause->getRequirements()) {
+            addRequirement(req, &protocolSubMap);
```
Comment (DougGregor) [inline lib/AST/GenericSignatureBuilder.cpp]: I *love* how compact this is!

File: lib/Sema/TypeCheckGeneric.cpp [hunk 8]
```diff
@@ -231,6 +267,67 @@ CompleteGenericTypeResolver::recordParamType(ParamDecl *decl, Type type) {
...
+      req.setInvalid();
+      return true;
+    }
+
+    if (validateType(req.getConstraintLoc(), lookupDC, options, resolver)) {
+      req.setInvalid();
+      return true;
+    }
+
+    // FIXME: Feels too early to perform this check.
```
Comment (DougGregor) [inline lib/Sema/TypeCheckGeneric.cpp]: I know this isn't your FIXME; this should eventually be handled by the generic signature builder, I think.

**Review by DougGregor — APPROVED** [review]
It's looking good!

**Comment by huonw** [comment]
@swift-ci Please smoke test.

**Comment by huonw** [comment]
@swift-ci Please smoke test

**Comment by huonw** [comment]
I think I've fixed everything, except for the rather ugly and unfortunate last two commits which just paper over bugs. I don't think I'll get time to address these until I get back, but it still might be useful to land this, for those depending on this feature.

**Comment by DougGregor** [comment]
We'll deal with the compiler crashers and get this merged

File: lib/AST/GenericSignature.cpp [hunk 9]
```diff
@@ -106,11 +106,11 @@ std::string GenericSignature::gatherGenericParamBindingsText(
     result += gp->getName().str();
     result += " = ";
 
-    auto found = substitutions.find(canonGP);
-    if (found == substitutions.end())
+    auto type = substitutions(canonGP);
```
Comment (slavapestov) [inline lib/AST/GenericSignature.cpp]: It is slightly more correct to do gp.subst(substitutions). Then you don't have to canonicalize it either. But feel free to merge this first and do this later

File: lib/Sema/TypeCheckGeneric.cpp [hunk 10]
```diff
@@ -128,6 +128,42 @@ void GenericTypeToArchetypeResolver::recordParamType(ParamDecl *decl, Type type)
         GenericEnv, type));
 }
 
+Type ProtocolRequirementTypeResolver::resolveGenericTypeParamType(
+    GenericTypeParamType *gp) {
+  assert(gp->getDepth() == 0 && gp->getIndex() == 0 &&
```
Comment (slavapestov) [inline lib/Sema/TypeCheckGeneric.cpp]: Self with non-zero depth does show up in invalid code though... In the past explicit checks like this have always caused problems eventually. You should change this to compare against Proto->getSelfInterfaceType() eventually

File: lib/Sema/TypeCheckGeneric.cpp [hunk 11]
```diff
@@ -128,6 +128,42 @@ void GenericTypeToArchetypeResolver::recordParamType(ParamDecl *decl, Type type)
...
+
+Type ProtocolRequirementTypeResolver::resolveDependentMemberType(
+    Type baseTy, DeclContext *DC, SourceRange baseRange,
+    ComponentIdentTypeRepr *ref) {
+  return DependentMemberType::get(baseTy, ref->getIdentifier());
+}
+
+Type ProtocolRequirementTypeResolver::resolveSelfAssociatedType(
+    Type selfTy, AssociatedTypeDecl *assocType) {
+  return assocType->getDeclaredInterfaceType();
```
Comment (slavapestov) [inline lib/Sema/TypeCheckGeneric.cpp]: Assert that 'self' is the protocol's self type then, if you're not substituting here?

**Comment by slavapestov** [comment]
@jrose-apple Do you think that's a bug though? If the RHS of the typealias were an associated type, and not concrete, we'd have to 'merge' the two associated types, ie:

    protocol ExpressibleByArrayLiteral { associatedtype Element }
    protocol Sequence { associatedtype Iterator : IteratorProtocol; typealias Element = Iterator.Element }
    protocol ArrayProtocol :  ExpressibleByArrayLiteral, Sequence {}

We really need to infer ExpressibleByArrayLiteral.Element == Sequence.Iterator.Element here.

Also, this example makes no sense at all:

```
protocol P {
  associatedtype A
  typealias B = A
}

protocol Q : P {
  typealias A = Int
}

extension P {
  func foo() {
    print(A.self)
    print(B.self)
  }
}

extension Q {
  func bar() {
    print(A.self)
    print(B.self)
  }
}

struct S1 : Q {
  typealias A = Float
}

print("Conformance of S1 to P: "); S1().foo()
print("Conformance of S1 to Q: "); S1().bar()
```

Note that typealias B is only defined once!

But inside the extension of P, 'B' resolves to the type witness for the associated type A, whereas inside the extension of Q, 'B' resolves to the type alias A, which is always Int.

I feel that the correct behavior is that inside a protocol, 'typealias T = ConcreteType' should always be equivalent to 'associatedtype T where T == ConcreteType', and 'typealias T = AssocType' should always be equivalent to  'associatedtype T where T == AssocType'.

We might be able to hack something up for Swift 3 mode, but it looks more of a consequence of bugs in the generic signature builder than anything intentional.

**Review by slavapestov — APPROVED** [review]

**Review by huonw — COMMENTED** [review]
I realised:

- I forgot to commit/push a fix I had for 3 of the crashers
- I don't know what's going on with the last one, IDE/021, sorry :(
- inherited protocols were crashing (but hopefully not, now)
- somewhat problematically same-type requirements between archetypes seem to disappear from the requirement signature: it looks like `enumerateRequirements` isn't passing them along. This means conformance checking doesn't catch bad things it should (i.e. if there's a `where A.T == B.T` in a protocol declaration, types can conform without this being satisfied), *but* the archetypes seem to still considered equal, so one can still use the types interchangeably in functions (see `needsInherited` in the last commit).

File: lib/AST/GenericSignature.cpp (hunk 9 above)
Comment (slavapestov) [inline lib/AST/GenericSignature.cpp]: You can pass in a MakeAbstractConformanceFn.

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by jrose-apple** [comment]
> I feel that the correct behavior is that inside a protocol, 'typealias T = ConcreteType' should always be equivalent to 'associatedtype T where T == ConcreteType', and 'typealias T = AssocType' should always be equivalent to 'associatedtype T where T == AssocType'.

That's a language change. The current behavior is just like anything defined in a protocol extension: it's shadowable and provides a default for associated types. (Okay, that's a bit of a stretch since typealiases in protocol extensions didn't work right in 3.0 IIRC. But I still don't think it's okay to change the meaning of this in Swift 3 mode.)

**Comment by DougGregor** [comment]
`requiresClassSlow()` isn't working properly due to the last commit. I'm testing a fix.

**Comment by DougGregor** [comment]
I believe that https://github.com/apple/swift/pull/7799 will address the failures in the standard library.

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
Cleanups to come, but I'd like to get this PR merged as-is to keep the history.

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test Linux

**Comment by DougGregor** [comment]
@swift-ci please smoke test Linux

**Comment by jrose-apple** [comment]
I'm sorry I missed this. I'm *very* terrified of `validateDeclForNameLookup`, because I don't think there's any guarantee that we'd come back and finish validating the decl later. What makes this safe?

**Comment by rjmccall** [comment]
Nothing. :)  I think that's what I fixed in the second commit of #7852.

**Pull request merged**